V2020.01 ts+ts7100

lionel opened the pull request.

Commit:
ts7100: Initial port from v2016 to v2020.01

- arch/arm/dts/Makefile
  - Add imx6ul-ts7100.dtb built from its dts, below
- arch/arm/dts/imx6ul-ts7100.dts
- arch/arm/mach-imx/mx6/Kconfig
  - Add config TARGET_TS7100 and
  - source "board/technologic/ts7100/Kconfig"

- configs/ts7100_defconfig (added file)
  - Leave CONFIG_SPL_BANNER_PRINT=y until board is in engineering sampling
  - Remove many unused configs & prefer defconfig over ts7100.h
    - e.g., `CONFIG_AUTOBOOT_KEYED` is in `configs/ts7100_defconfig` rather
      than `include/configs/ts7100.h`.
    - CONFIG_BOOTDELAY 10 -> 5 seconds
    - Also, we don't want `ts7100.h` to hardcode "run silochargeon" into
      `CONFIG_PREBOOT` when it is available to be added to the bootcmd.
    - CONFIG_BOARD_EARLY_INIT_F=y
    - CONFIG_MXC_SPI=y

- Added board/technologic/ts7100/ subdirectory with:
  - Makefile
  - Kconfig
  - fram.c
  - fram.h
  - led.c
  - mfg.c
  - parse_strap.c
  - parse_strap.h
  - spl.c
    - memory sizing for 512 MiB or 1 GiB based on strapping
  - ts7100.c
    - Call to is_mfg() overrides bootcmd (to mfg command) and bootdelay.
  - silabs.c
	printf->debug
        #undef DEBUG
  - tsfpga.h

- include/configs/ts7100.h (added file)
  - Linux fdtfile hardcoded to imx6ul-ts7100-0.dtb
  - distroboot support present in `include/configs/ts7100.h` but not working.

Versus v2016, CONFIG_OCSPI_BASE is trimmed down to OCSPI_BASE to
avoid Kconfig conflict and build complaints.

lionel committed:
ts7100: Rev A strapping resistor support

lionel committed:
ts7100: Using the watchdog to reset_cpu keeps console from dropping

lionel committed:
ts7100: mfg: remove duplicate call of mmc_set_boot_bus_width

Inherited from the ts7250v3 (fix there too)

lionel committed:
ts7100: dts: Fix PHY numbering & document how it pieces together

PHY naming/numbering is a gigantic maze of inconsistencies. No wonder
we had it wrong.

lionel committed:
ts7100: Rename of TS-7100 names and URLs to embeddedTS

lionel committed:
ts7100: mfg: Fix access to 9471 (display of priming step's outcome)

lionel committed:
ts7100: For boot areas down to 2MiB, reloc environment to 0x100000

This change impacts `/etc/fw_env.config` and everything from priming
to blasting, buildroots, and debians.

lionel committed:
ts7100: dts: Fix gpio-line-names

lionel committed:
ts7100: Rev B & custom strapping resistor support plus FPGA_STRAPS

Strapping code is partitioned out so that future strapping updates
should not need to touch ts7100.c.

Distinct `read_raw_*` strap functions handle the reading and caching
of pad values. Besides masking out non-straps, the only processing
performed is inverting bits for pull-downs as necessary so that:

  1=POP/present and 0-DEPOP/not-preset/NA

The `read_{cpu,io}_board_*` functions return strap register values for
`opts`, `io_model`, and `io_opts`. Bits follow a common arrangement
that specifies every TS-7100-family combination/variant.

The remaining `get_*` functions return strings based on interpreting
the FPGA and CPU straps.

lionel committed:
ts7100: Change default env to align with current distroboot practice

* emmcboot and usbprod are no longer the way to boot from emmc and usb
* distroboot script expects fdtfile set based on io_model
* distro-seed's distroboot script doesn't want cmdline_append?
* distro-seed's distroboot script wants console in env?

fixup! ts7100: distro-seed's distroboot script wants console in env?

lionel committed:
FIXME: ts7100: lower bootdelay from 5 seconds to 1

This is a temporary "common" solution as we migrate to a different
(ideally 0) bootdelay when no console is attached.

lionel committed:
FIXME: ts7100: With FPGA > v6, save straps for userspace tshwctl to read

We are using intentionally wrong FPGA revs until the v6 is tested.

markfeathers commented on board/technologic/ts7100/parse_strap.c:
```diff
@@ -0,0 +1,225 @@
+  return "TS-7100";
+}
+
+const char *get_cpu_board_version(void)
+{
+	uint8_t io_model;
+	uint8_t cpu_board_rev;
+        uint16_t raw_cpu_straps;
+        uint16_t raw_fpga_straps;
+	static char model_str[16] = {0};
+
+	io_model = read_io_board_model();
+        raw_cpu_straps =  read_raw_cpu_straps();
+        raw_fpga_straps =  read_raw_fpga_straps();
+
+        cpu_board_rev = ((raw_cpu_straps >> 7) & 0x02);    // Rev B 0x02 strap on the 6UL
+        cpu_board_rev |= ((raw_fpga_straps >> 12) & 0x01); // Rev A 0x01 strap on the FPGA
+
+        if (cpu_board_rev & 0x2) {
+          snprintf(model_str, sizeof(model_str), "B");
+          return model_str;
+        }
+        if (cpu_board_rev & 0x1) {
+          snprintf(model_str, sizeof(model_str), "A");
+          return model_str;
+        }
+        if (io_model == 0) {
+          snprintf(model_str, sizeof(model_str), "P2-%02x", cpu_board_rev);
+        } else if (io_model == 1) {
+          return "Am";
+        } else if (io_model == 3) {
+          snprintf(model_str, sizeof(model_str), "B-%02x", cpu_board_rev);
+        } else {
+          snprintf(model_str, sizeof(model_str), "UNKNOWN_%02x", cpu_board_rev);
+        }
+        return model_str;
```
We talked about this one a little bit yesterday, but I'm going to suggest we do something like this:

```suggestion
const char get_cpu_board_version(void)
{
        uint8_t rev_b_strap;
        
	io_model = read_io_board_model();
        raw_cpu_straps =  read_raw_cpu_straps();
        rev_b_strap =  = ((raw_cpu_straps >> 7) & 0x02);

        if (rev_b_strap) {
                return 'B';
        } else {
                return 'A';
        }
}
```

This makes it a bit cheaper/simpler to do the rev comparisons when we need to instead of doing full string operations, eg:

```
if(get_cpu_board_version() == 'A') {
...
```

lionel replied:
Since putting up this code, I modified it to do basically what you're saying.

The assumption is we'll check bits from newest to oldest known hardware and return immediately when there is a "hit". Boards newer than rev B **should** return rev B.

I do check the "`rev_a_strap`" and fall-through to handle cases where something doesn't "add up". It should never be reached on hardware we release, but is a reality where speedy debug-ability has value. Meaning, a valid rev "A" should not be returned for something that is definitely wrong. If it is newer than B, we've already said it should return B, not A. If there is a mistake, we should make it easy to detect and track down.

< Lionel

markfeathers commented on board/technologic/ts7100/parse_strap.c:
```diff
@@ -0,0 +1,225 @@
+
+uint8_t read_io_board_model(void)
+{
+  uint8_t io_model;
```
I dont know that we need to meet checkpatch on this u-boot version, but make sure the whitespace is consistent through the file.  In general u-boot should follow the same standards as linux using 8 character tabs.

lionel replied:
Thank you.  Will fix. I don't know how some of my machines' editor settings got messed up, but will fix that too.

markfeathers commented on board/technologic/ts7100/parse_strap.c:
```diff
@@ -0,0 +1,225 @@
+  return io_opts;
+}
+
+/* NOTE: Before this function is called, it expects that the IOMUX be set up
+ * ahead of time. Meaning, pin function (GPIO) and pullup need to be set up
+ * before the read_raw_cpu_straps() function is called. (FIX WORDING)
+ */
```
Why not do that in this function or a light wrapper around this function?  If this is handled nearby or at the same time this might be simpler as we need to add new straps in the future.

lionel replied:
Good point. I split out the straps from `misc_pads` (so they aren't needlessly coupled with `ts7100.c`) and put them in `parse_strap.c` as `strap_pads`.

markfeathers commented on board/technologic/ts7100/spl.c:
```diff
@@ -0,0 +1,301 @@
+
+	/* Wait for sane fpga, should take 62ms */
+	for (i = 0; i < 1000; i++) {
+#if 0
+		if(!gpio_get_value(FPGA_RESET_GPIO))
+#else
+                  if ((readl(FPGA_SYSCON + 0x50) & 0x2) == 0)
+#endif
```
```suggestion
                  if ((readl(FPGA_SYSCON + 0x50) & 0x2) == 0)
```

lionel replied:
Fixed.

markfeathers commented on board/technologic/ts7100/spl.c:
```diff
@@ -0,0 +1,301 @@
+	.dram_reset = 0x00000030,
+};
+
+static struct mx6_mmdc_calibration ts7250v3_512m_calibration = {
```
```suggestion
static struct mx6_mmdc_calibration ts7100_512m_calibration = {
```

lionel replied:
I had planned to leave this text until the 7100 is really calibrated and not cribbing off the 7250v3 !-)

markfeathers commented on board/technologic/ts7100/spl.c:
```diff
@@ -0,0 +1,301 @@
+};
+#  endif
+#endif
+struct mx6_ddr_sysinfo ts7250v3_sysinfo = {
```
```suggestion
struct mx6_ddr_sysinfo ts7100_sysinfo = {
```

markfeathers commented on board/technologic/ts7100/spl.c:
```diff
@@ -0,0 +1,301 @@
+          /* 8Gb Alliance AS4C512M16D3L-12BCN */
+          mx6_dram_cfg(&ts7250v3_sysinfo,
+                       &ts7250v3_512m_calibration,
+                       &ts7250v3_1g_ddr);
```
All of the ts7250v3 references should be updated in this file, but we should definitely be using a sane 1g calibration for the 8gb chipset.

markfeathers commented on board/technologic/ts7100/parse_strap.c:
```diff
@@ -0,0 +1,225 @@
+
+		gpio_direction_input(JTAG_TDO_N15);
+
+		mdelay(1);
```
We do not need a sleep in between these. Realistically the CPU pins are in the 1-2 MHz ballpark for reads. It will take ~20-60ns for the pins to transition from vil/vih from their pulls, so the ~500ns period of an access alone is plenty. FWIW the fastest GPIO we typically see on an SOC is ~10mhz so 100ns would still be plenty for a typical pull.

markfeathers commented on board/technologic/ts7100/parse_strap.c:
```diff
@@ -0,0 +1,225 @@
+                  saved_straps = readl(FPGA_STRAPS);
+                  saved_straps |= (fpga_straps << 16);
+                  writel(saved_straps, FPGA_STRAPS);
+                }
```
At this stage in its lifetime I'd advise against trying to support all combinations of FPGA/u-boot/PCB. This should be saved for once we're out of engineering sampling before we accept added complexity for all combinations.  We can require users to update FPGAs and u-boot on these products to correctly detect the revisions using the new method that will be consistent going forward.

This code should assume all boards support the new straps register.

markfeathers commented on board/technologic/ts7100/parse_strap.c:
```diff
@@ -0,0 +1,225 @@
+    return "TS-7100-Z";
+  }
+  return "TS-7100";
+}
```
For now I'd recommend we only return "TS-7100" here. In general in u-boot scripting this is used to discern product families.  Eg, so we can pick how a TS-7100 loads an FPGA/initramfs/u-boot/etc which are different in some respects than a TS-7250-V3 or a TS-4100. How the TS-7100 works is the same between carrier boards so we shouldn't try to match further than that. We further identify carrier boards through the other id variables if needed, or a more friendly name once we get to linux with something like /proc/device-tree/model.

markfeathers commented on board/technologic/ts7100/ts7100.c:
```diff
@@ -0,0 +1,620 @@
+	gpio_direction_input(JTAG_FPGA_TMS);
+	gpio_direction_input(JTAG_FPGA_TDO);
+
+	/* During FPGA programming several important pins will
+	 * have been tristated.  Put it back to normal */
+	//fpga_mmc_init();
+	//red_led_on();
+	//green_led_off();
+	//fpga_gpio_output(OFF_BD_RESET_PADN, 1);
+	//fpga_gpio_output(EN_USB_HOST_5V_PAD, 1);
```
I think we can remove these.  If we want to talk about supporting soft/ram reloaded fpgas we should set sane defaults in the fpga itself.

markfeathers commented on include/configs/ts7100.h:
```diff
@@ -0,0 +1,224 @@
+	"initrd_high=0xffffffff\0" \
+	"autoload=no\0" \
+	"clearenv=env default -f -a && env save;\0" \
+	"emmc_dev=0\0" \
```
```suggestion
```
Not used as far as I can see, should be removed.

lionel replied:
Useful for scripts that want to know whether eMMC is `mmcblk0` or `mmcblk1`.

lionel replied:
And, BTW, the TS-7250v3 has it too !-)

lionel replied:
Removed.

markfeathers commented on include/configs/ts7100.h:
```diff
@@ -0,0 +1,224 @@
+			"else scaps wait pct ${chrg_pct};" \
+		"fi;" \
+	"fi;\0" \
+	"usbprod=usb start;" \
+		"if usb storage;" \
+			"then echo Checking USB storage for updates;" \
+			"if load usb 0:1 ${loadaddr} /tsinit.scr;" \
+				"then led green on;" \
+				"source ${loadaddr};" \
+				"led red off;" \
+				"exit;" \
+			"fi;" \
+			"if load usb 0:1 ${loadaddr} /tsinit.scr.uimg;" \
+				"then led green on;" \
+				"source ${loadaddr};" \
+				"led red off;" \
+				"exit;" \
+			"fi;" \
+		"fi;\0" \
+	"emmcboot=echo Booting from the eMMC ...;" \
+		"if load mmc 0:1 ${loadaddr} /boot/boot.ub;" \
+			"then echo Booting from custom /boot/boot.ub;" \
+			"source ${loadaddr};" \
+		"fi;" \
+		"load mmc 0:1 ${fdtaddr} " \
+		  "/boot/imx6ul-ts${model}-${io_model}.dtb;" \
+		"if load mmc 0:1 ${loadaddr} /boot/zImage;" \
+			"then run silowaitcharge;" \
+			"setenv bootargs root=/dev/mmcblk0p1 rootwait rw " \
+			  "cpu_opts=0x${opts} io_opts=0x${io_opts} " \
+			  "io_model=0x${io_model} ${cmdline_append};" \
+			"bootz ${loadaddr} - ${fdtaddr};" \
+		"else echo Failed to load kernel from eMMC;" \
+		"fi;\0" \
```
```suggestion
```

These are from the 2016 u-boot an are deprecated with distroboot.

lionel replied:
The latter has been super-useful, particularly when distroboot breaks. Or a direct boot is wanted. If removed, I think it better to do so under a separate commit that can be put back.

lionel replied:
Also, there are many other boards still using `emmcboot`, such as the Colibris. One (first in the list below) actually hedges its bets and uses distroboot as a fallback:
```
$ git grep emmcboot  | grep '[^ ]=[^ ]'
configs/apalis-tk1_defconfig:CONFIG_BOOTCOMMAND="run emmcboot; setenv fdtfile ${soc}-${fdt-module}-${fdt_board}.dtb && run distro_bootcmd"
include/configs/apalis-tk1.h:	"emmcboot=run setup; run emmcfinduuid; run set_emmcargs; " \
include/configs/apalis-tk1.h:	"emmcbootpart=1\0" \
include/configs/apalis-tk1.h:	"emmcdtbload=setenv dtbparam; load mmc ${emmcdev}:${emmcbootpart} " \
include/configs/apalis_imx6.h:	"emmcboot=run setup; run emmcfinduuid; run set_emmcargs; " \
include/configs/apalis_imx6.h:	"emmcbootpart=1\0" \
include/configs/apalis_imx6.h:	"emmcdtbload=setenv dtbparam; load mmc ${emmcdev}:${emmcbootpart} " \
include/configs/cl-som-imx7.h:	"emmcbootscript=setenv mmcdev 1; setenv mmcblk 2; run mmcbootscript\0" \
include/configs/cl-som-imx7.h:	"emmcboot=setenv mmcdev 1; setenv mmcblk 2; run mmcboot\0" \
include/configs/cm_t43.h:	"emmcboot=echo Booting from emmc ... && " \
include/configs/colibri_imx6.h:	"emmcboot=run setup; run emmcfinduuid; run set_emmcargs; " \
include/configs/colibri_imx6.h:	"emmcbootpart=1\0" \
include/configs/colibri_imx6.h:	"emmcdtbload=setenv dtbparam; load mmc ${emmcdev}:${emmcbootpart} " \
include/configs/colibri_imx7.h:	"emmcboot=run setup; run emmcfinduuid; run set_emmcargs; " \
include/configs/colibri_imx7.h:	"emmcbootpart=1\0" \
include/configs/nokia_rx51.h:	"emmcboot=setenv mmcnum 1; run trymmcboot\0" \
include/configs/nokia_rx51.h:	"bootmenu_1=Internal eMMC=run emmcboot\0" \
include/configs/opos6uldev.h:	"emmcboot=run initargs; run addmmcargs; "                                               \
include/configs/opos6uldev.h:	"emmcsafeboot=setenv mmcpart 1; setenv mmcroot /dev/mmcblk0p1 ro; run emmcboot;\0"      \
include/configs/uniphier.h:	"emmcboot=mmcsetn && run bootcmd_mmc${mmc_first_dev}\0" \
$ 
```

lionel replied:
Something else worth noting: Many other platforms (n=332 defconfigs), particularly those omitting `emmcboot`, have `CONFIG_DISTRO_DEFAULTS` in their defconfigs. This config is mentioned in `doc/README.distro`.

lionel replied:
Removed as suggested.

markfeathers commented on include/configs/ts7100.h:
```diff
@@ -0,0 +1,224 @@
+		"fi;\0" \
+        BOOTENV
+
+#define UPDATER_ENV \
+	"update-spl=dhcp;"\
+		"if nfs ${loadaddr} ${nfsip}:${nfsroot}/boot/SPL; then " \
+                        "setexpr filesize ${filesize} / 200 ; " \
+                        "setexpr filesize ${filesize} + 1 ; " \
+                        "mmc dev 0 1 ; " \
+                        "mmc write ${loadaddr} 2 ${filesize} ; "\
+                "fi;\0" \
+        "u-boot-bin=u-boot-dtb.img\0" \
+        "u-boot-dir=boot/\0" \
+	"u-boot-flash=setexpr filesize ${filesize} / 200 ; " \
+                "setexpr filesize ${filesize} + 1 ; " \
+                "mmc dev 0 1 ; " \
+                "mmc write ${loadaddr} 8a ${filesize} ;\0" \
+	"update-uboot=dhcp;"\
+		"if nfs ${loadaddr} ${nfsip}:${nfsroot}/${u-boot-dir}${u-boot-bin}; then " \
+                        "setexpr filesize ${filesize} / 200 ; " \
+                        "setexpr filesize ${filesize} + 1 ; " \
+                        "mmc dev 0 1;" \
+                        "mmc write ${loadaddr} 8a ${filesize};"\
+                "fi;\0"
```
```suggestion
```

This appears unused and should not be included.  This might be better as separate internal documentation.

lionel replied:
Removed.

markfeathers commented on lib/panic.c:
```diff
@@ -31,6 +31,7 @@ static void panic_finish(void)
 
 void panic_str(const char *str)
 {
+	puts("PANIC: ");
```
```suggestion
```

lionel replied:
Pulled out into my debugging branch.

markfeathers commented on scripts/config_whitelist.txt:
```diff
@@ -556,7 +556,6 @@ CONFIG_FORCE_DDR_DATA_BUS_WIDTH_32
 CONFIG_FORMIKE
 CONFIG_FPGA_COUNT
 CONFIG_FPGA_DELAY
-CONFIG_FPGA_LATTICE
```
I'm not sure I understand this change, why remove this from the whitelist?

lionel replied:
I don't know !-) lol When removing the FPGA programming way back when, it must have seemed it was needed for some reason. I put it back.

luckshot replied:
I am curious:  What is CONFIG_FORMIKE supposed to do?

lionel replied:
@luckshot : In `drivers/video`, it includes a module called `drivers/video/formike.c`. See commit `b26354`.

lionel committed:
ts7100: Add FPGA revision display to checkboard()

lionel committed:
ts7100: bootcmd_mfg is subsumed by the mfg command

lionel committed:
ts7100: Remove FRAM and its use for BOOTCOUNT storage

FRAM is incompatible with how Mender uses bootcount, so is being
removed.

FRAM isn't large enough to store a U-Boot environment, and that is where
bootcount is normally tracked.

lionel committed:
ts7100: Remove emmc_dev from default environment

markfeathers commented on configs/ts7100_defconfig:
```diff
@@ -0,0 +1,134 @@
+CONFIG_PANIC_HANG=y
+CONFIG_SPL_RAW_IMAGE_SUPPORT=y
+CONFIG_SPL_LIBDISK_SUPPORT=y
+CONFIG_SMBIOS_PRODUCT_NAME="mx6ul_14x14_evk"
```
```suggestion
CONFIG_SMBIOS_PRODUCT_NAME=""
```
I'd normally say we should correct it rather than remove it, but this is gone on newer u-boots and making it blank otherwise is a normal convention:
https://patchwork.ozlabs.org/project/uboot/patch/20201022141949.711984-13-sjg@chromium.org/

lionel replied:
It looks like that is the convention for boards that actually use/support SMBIOS. Most defconfigs omit SMBIOS entirely, so I deleted it from the 7100 (and 7250v3), the only boards of ours that had it.

markfeathers commented on configs/ts7100_defconfig:
```diff
@@ -0,0 +1,134 @@
+CONFIG_SYS_WHITE_ON_BLACK=y
+CONFIG_FS_BTRFS=y
+CONFIG_FS_CBFS=y
+CONFIG_SMBIOS_MANUFACTURER="freescale"
```
```suggestion
CONFIG_SMBIOS_MANUFACTURER=""
```

Same as above

lionel committed:
ts7250v3: Remove emmc_dev from default environment

lionel committed:
ts7100: spl: Update DDR3 to latest calibration

lionel committed:
ts7250v3: Remove SMBIOS from defconfig

Only 17 defconfigs besides ours have them, most empty. SMB booting is
not applicable to our products anyway.

lionel committed:
ts7100: Remove SMBIOS from defconfig

Only 17 defconfigs besides ours have them, most empty. SMB booting is
not applicable to our products anyway.

lionel committed:
ts7100: WDOG reset used by 5.10 kernel alters FPGA startup timing

After a CPU WDOG# reset (as opposed to a Silabs watchdog reset), the
DRAM sizing strap on the FPGA reads as 1GB for 512MB boards.

Reads OK with 4.9 and U-Boot resets, as well as from power-on.

This commit is associated with FPGA rev 23. It uses the FPGA_INT2 GPIO
to signal the CPU when it is finished with its reset. Up to 45 ms
will be given for the FPGA to drive FPGA_INT2 low.

As of this commit, only 5 mdelay calls are needed before the FPGA
reports sane and gives the correct DRAM strapping.
